fix: use screen for reconnecting terminal sessions if available

code-asher opened the pull request:
This fixes https://github.com/coder/coder/issues/3666 and similar issues.

I copied much of this from v1 (https://github.com/coder/wsep/pull/25) but one major difference is that I was asked to keep the existing behavior as a fallback so I broke out parts into an interface then moved the existing code into said interface.

I think having a fallback makes more sense than it did for v1 since we have macOS and Windows that are not using screen yet (I have not looked into whether screen can run on Windows yet and on macOS it works for me locally but in CI screen will sometimes refuse to find a session even if you `screen -list` to verify that exact name does in fact exist).

code-asher committed:
Add screen backend for reconnecting ptys

The screen portion is a port from wsep.  There is an interface that lets
you choose between screen and the previous method.  By default it will
choose screen if it is installed but this can be overidden (mostly for
tests).

The tests use a scanner instead of a reader now because the reader will
loop infinitely at the end of a stream.

Relpace /bin/bash with bash since bash is not always in /bin.

spikecurtis commented on agent/reconnectingpty/buffered.go:
```diff
@@ -0,0 +1,167 @@
+		b.metrics.WithLabelValues("write").Add(1)
+		return nil, xerrors.Errorf("write buffer to conn: %w", err)
+	}
+	b.activeConns[connID] = conn
```
there is a race condition here:  the command may have already exited and run its routine that closes all connections.  Then, we come in and add a new connection, which will just hang forever if there is no input.

We need to store lifecycle information that we can check while holding a mutex to decide whether the pty is still open.

code-asher replied:
I think I could also fix this if instead of using a map I used channels (send pty output to a channel, select on that in `attach` and write to the conn).

But even if that works, do you think tracking the state in the upper layer is still weird enough that I should still move it to the backend?  Lifecycle is almost entirely what the upper layer does so I am wondering if I just get rid of that layer entirely.

Hmmmm yeah I think I will do that, the backend really should understand its own state even if this particular bug goes away, but let me know if this seems wrong.  Will probably just keep `New` as-is and move the rest to the individual backends.

code-asher replied:
Refactored!  I think I was prematurely abstracting before.  There is some duplication we might want to examine but I think overall things make more sense now.

spikecurtis commented on agent/reconnectingpty/reconnectingpty.go:
```diff
@@ -0,0 +1,298 @@
+type backend interface {
+	start(ctx context.Context, logger slog.Logger) error
+	attach(ctx context.Context, connID string, conn net.Conn, height, width uint16, logger slog.Logger) (pty.PTYCmd, error)
+	close(ctx context.Context, logger slog.Logger) error
```
the use of `error` on `close()` seems fairly arbitrary in this code.  The buffered backend returns errors closing connections but not errors closing the pty or killing the process.

Considering that the only caller of `close()` just logs the error anyway, I wonder whether it would be simpler to not return an error and have the backend be responsible for logging any problems.

code-asher replied:
Ahh good point, I will remove the return.

I think also I was too lazy here; the process might already have exited so I did not want to return that error.  I will properly fix that along with the `cleanup` fix you mentioned above.

code-asher committed:
Fix leaking goroutine in wait

code-asher commented on agent/reconnectingpty/reconnectingpty.go:
```diff
@@ -0,0 +1,298 @@
+
+type backend interface {
+	start(ctx context.Context, logger slog.Logger) error
+	attach(ctx context.Context, connID string, conn net.Conn, height, width uint16, logger slog.Logger) (pty.PTYCmd, error)
```
Really good points, will refactor.  I originally approached it as just breaking out the parts that were different between the two but I think reapproaching the design is the better option.

code-asher committed:
Remove connection_id from reconnecting PTY

This serves multiple connections so it makes no sense to scope it to a
single connection.

Also lets us use "connection_id" when logging write errors instead of
"other_conn_id".

code-asher commented on agent/agent.go:
```diff
@@ -1119,169 +1115,25 @@ func (a *agent) handleReconnectingPTY(ctx context.Context, logger slog.Logger, m
+			BackendType: msg.BackendType,
+			Timeout:     a.reconnectingPTYTimeout,
+			Metrics:     a.metrics.reconnectingPTYErrors,
+		}, logger)
```
Ahh good point, I left it like this to keep the previous behavior but I should have just fixed it.  Will do!

code-asher commented on agent/reconnectingpty/reconnectingpty.go:
```diff
@@ -0,0 +1,298 @@
+	options *Options
+	// state holds the current reconnecting pty state.  It is not safe to access
+	// this outside of cond.L.
+	state State
```
Ahh yup I see what you mean.  Gonna refactor.

code-asher commented on codersdk/workspaceagentconn.go:
```diff
@@ -193,13 +193,32 @@ func (c *WorkspaceAgentConn) Close() error {
+	Height      uint16
+	Width       uint16
+	Command     string
+	BackendType ReconnectingPTYBackendType
```
When I added this I had the same thought.  It does fall back if you set nothing (or set `auto`) but the main reason I added it was for tests.

It might be enough that we test both backends as a consequence of running tests on multiple platforms, but it feels a bit sketchy to me.

An idea I had was to test the reconnecting pty directly rather than through the SDK as we currently do, but I vaguely remember hearing that pattern was discouraged (it is possible I have misremembered something).

spikecurtis replied:
Is it possible to manipulate the test environments so that `screen` is available in one test case, but not available in another?  That would also allow us to test the code that falls back to the circular buffer.

code-asher replied:
I had a somewhat janky idea to symlink bash into a directory and set `PATH` to that so screen is not available which does work.  I will push that up for now, but maybe there is a more clever way of doing this.

code-asher committed:
Remove error from close return

Some errors are returned and some are logged at the debug level.  The
reason for the debug logging is that the process might already be closed
so in that case I did not want to return an error.  This will be fixed
in a future commit but since this error is only logged by the caller
anyway, go ahead and just log it in the backend to simplify things.

code-asher committed:
Refactor reconnecting PTY backends

Instead of a shared layer, directly implement both and share code where
possible with helper functions.

This makes cleanup easier with the screen implementation too.

code-asher committed:
Merge remote-tracking branch 'github/main' into asher/reconnection-with-screen

code-asher committed:
Remove extra mutex unlock

code-asher committed:
Fix heartbeat typo in comment

code-asher committed:
Tweak connection close

We need to close the connection when the process exits otherwise the
read will keep blocking.

No need to close it where we were before since the agent already closes
the connection once Attach yields.

code-asher committed:
Linter fixes

I missed an err, it dislikes the attach/Attach naming, and some unused
variables.

spikecurtis commented on agent/reconnectingpty/buffered.go:
```diff
@@ -0,0 +1,238 @@
+
+	// Pipe conn -> pty and block.  pty -> conn is handled in newBuffered().
+	readConnLoop(ctx, conn, rpty.ptty, rpty.metrics, logger)
+	return nil
```
Your comment on `Attach()` says:

> If the context ends or the process dies the connection will be detached and closed.

But I don't think that it really true.

Firstly, if the `ctx` passed in ends after we are `Ready`, I don't think there is any direct mechanism to detect that.  `readConnLoop` only uses the context for logging; it doesn't directly check whether the context expires.

Now, you might think: well, if the context expires, then the `lifecycle()` will close the connection, which will trigger an EOF when `readConnLoop` reads from it.  But!, that's the context passed to `newBuffered()`, **not** the context passed to `Attach()`.  As it happens, the agent uses the same context for both, but if we're depending on that behavior, then `Attach()` shouldn't accept a context, and we should just save the context passed to `newBuffered()` on the struct.

Secondly, if we go through the deferred function on L184, then the connection will not be closed.  This can happen if the `ptty` dies and `readConnLoop` returns before the `lifecycle()` can lock the mutex.

I don't think it's strictly necessary for `Attach()` to close the connections, since `handleReconnectingPTY` closes the conn in a defer, but we shouldn't claim on the interface that we will close it.

code-asher replied:
I tweaked things to close the connection in `Attach()` to fix the issue with adding to the map after closing, and I went ahead and did it on context end as well so if the caller cancels their context then we break out of the loop. 

But having to close the connection to make `readConnLoop` yield is I think just an implementation detail and it still makes sense to avoid claiming `Attach()` closes the connection, so I left that part out of the comment!

Maybe it would be better to handle the context end in `readConnLoop` explicitly instead of relying on closing the connection to terminate that loop though, and this would avoid the potentially useless debug log when a connection is already closed.

code-asher replied:
Ah nevermind I am not seeing a way to make the connection read abort when the context ends, so I guess we have to close it to make it yield.

spikecurtis replied:
If you want to Read, but also abort on context you can do your reading in a separate goroutine, and send the results (read data and errors) over channels.  In the main goroutine you `select` on  `<-context.Done()` or the results channels.

The tricky bit is that you also need to wrap the sending on channels in a `select` that includes `<-context.Done()` so that you don't leak the Reader goroutine if the context ends and the main goroutine exits.

That's a lot of complexity, so I'm fine with closing the channel...

code-asher replied:
Ahhhhhhh I see what I missed, I was too fixated on making the read itself abort, but if we have the read in a goroutine then it would not block `readConnLoop`, and once we yield since the agent closes the connection it would then make the read yield and the read goroutine would exit.

I will go ahead and keep closing it for now, but thank you for the explanation!

spikecurtis commented on agent/reconnectingpty/reconnectingpty.go:
```diff
@@ -0,0 +1,227 @@
+	Wait()
+	// Close kills the reconnecting pty process.
+	Close(reason string)
+}
```
This interface is a big improvement.  I like it.

code-asher replied:
Thank you!!

spikecurtis reviewed (commented):
This is a big improvement, just a few lingering races in comments, and some commentary about clear interfaces.

code-asher commented on agent/reconnectingpty/buffered.go:
```diff
@@ -0,0 +1,238 @@
+	rpty.mutex.Lock()
+	defer rpty.mutex.Unlock()
+
+	state, err := rpty.state.waitForStateOrContext(ctx, StateReady)
```
This makes sense and I had the same concerns.  My thinking though is that I would need to hold the mutex while checking the state otherwise if the pty closes in the meantime I will end up adding it to the map and the conn might not get closed.

Assuming my thinking is correct, maybe this means I should refactor to avoid using a map in the first place.

code-asher replied:
Actually I think I can fix this by adding a new goroutine that closes the conn when the state changes to closed.

spikecurtis replied:
That's a good point.  If you keep the map, it indicates to me that there should only be one lock, not two.  Since you should only insert into the map if the reconnectingPTY is in a specific state, both the map and the state should be covered by a single lock.  That also dodges issues of deadlocking, and since waiting for a `Cond` releases the lock while you are waiting, it also avoids the performance issues.

code-asher replied:
Great idea!  I implemented this with a function passed to `waitForStateOrContext()` so I could manipulate the map in that function while holding the `Cond` lock.  It feels a bit JavaScript-y to me and I am not sure if it looks like idiomatic Go so please let me know if you think it looks weird.

code-asher commented on agent/reconnectingpty/buffered.go:
```diff
@@ -0,0 +1,238 @@
+		if err != nil {
+			logger.Debug(ctx, "closed conn with error", slog.Error(err))
+		}
+	}
```
Great point!

code-asher committed:
Clear active conns on close

code-asher committed:
Avoid useless buffer reset on close

code-asher commented on agent/reconnectingpty/buffered.go:
```diff
@@ -0,0 +1,238 @@
+		timeout:     options.Timeout,
+	}
+
+	go rpty.lifecycle(ctx, logger)
```
Oof I should have caught that.  Thank you.

code-asher committed:
Move lifecycle after buffer and process are set

The lifecycle will try to close the process (and before it would clear
the buffer as well) so they need to be set.

code-asher committed:
Add info logs for starting, stopping, and attaching

code-asher committed:
Do not hold mutex while waiting for state in screen

Holding off on changing this for the buffered version until a potential
issue with the pty closing before we add to the map is sorted.

I moved the heartbeat start out as well since while we want to wait to
start that until after ready, it does not need to be behind the mutex.

code-asher committed:
Remove incorrect statement about closing on Attach

In some cases it might get closed but it is not a guarantee like the
comment claims it is.

While we could implement this behavior properly we already close the
connection in the caller so it seems unnecessary.

code-asher committed:
Remove backend type from SDK/API

Still need to figure out a way to test both, but we would rather not add
the ability to select the backend through the API.

code-asher committed:
Use PATH to test buffered reconnecting pty

code-asher committed:
Do not hold mutex while waiting for state

Since we might add to the map after it closes, tweak the delete to also
run when the pty closes.

I think this also fixes that issue where if you canceled the context you
pass into Attach it would not do anything since readConnLoop does not
use the context for anything but logging.

code-asher committed:
Avoid clobbering attach error with close errors

We want to send the error from failing to wait for the version command,
not the error from closing/killing the pty or process.

code-asher committed:
Immediately read screen process

It is possible for the process to immediately exit (for example if you
run `echo hello`), then we would wait for the `version` command to
succeed but it never will because the session is already gone.

If we immediately read to the process then we can tell when it has gone
away and we can abort.

code-asher committed:
Fix incorrect logger context on reconnecting PTY

Since we used the same var all I did with my other change was use the
same logger anyway.

Now it will not have the connection ID, which is correct since the
pty is not tied to a single connection.

spikecurtis reviewed (approved):
LGTM.

I do think it would be an improvement to consolidate the bufferedReconnectingPTY down to a single lock, as I mentioned in a comment thread, but what you have now is fine.

code-asher committed:
Protect map and state with the same mutex

I moved the conn closes back to the lifecycle, too.

code-asher committed:
Fix incorrect test comment

code-asher committed:
Attempt fixing flake with 'echo test' command

code-asher commented on agent/reconnectingpty/buffered.go:
```diff
@@ -0,0 +1,234 @@
+	defer cancel()
+
+	// Once we are ready, attach the active connection while we hold the mutex.
+	_, err := rpty.state.waitForStateOrContext(ctx, StateReady, func(state State, err error) error {
```
It does feel odd.

Mostly I needed to write out the buffer even if the state is closed to fix a flake where the pty closes too quickly when there is a short-lived command, and then after that I check the state and return the original `err`.

Really this will only happen in tests though, I am not sure why we spawn things like `echo hello` when real-world usage is always spawning a shell.

I could just duplicate writing out the buffer outside the wait, before returning the error.

In the case of context expiring, I can skip the function call, this is only really needed for the quickly-closing case.  I will go ahead and do that now, and let me know if it would be smarter to duplicate the buffer write!

Somewhat related, I wonder sometimes if I should rename `error` to `reason`, since it is almost more like a reason for the state change.

code-asher replied:
Or I could actually still omit the error and just have `waitForStateOrContext` return the original error if `fn` returns nil.

spikecurtis replied:
It's starting to feel awkward to pass in the function because the circumstances under which you do and do not want to execute the function are hard to express.  Another option to consider: split `waitForStateOrContext` into a `waitForStateOrContextLocked` function that does the stuff assuming the caller has locked the `cond`.  Then, you can hold the lock here, call `waitForStateOrContextLocked`, then do your extra processing before releasing the lock.

code-asher replied:
That makes sense, and feels cleaner.  Will fix.

code-asher committed:
Avoid wait callback when context expires

code-asher committed:
Remove err from wait callback

We do not really do anything with it other than just return it, and if
we do need to do something with it we can just do it on the return from
`waitForStateOrContext`, we probably would not need to use it behind the
mutex.

Also we should check the state outside since there is technically no
guarantee an error is set on a state change (although in practice for
close/done there always is, maybe this should be enforced in some way).

code-asher committed:
Add state wait func where caller holds the lock

code-asher committed:
Merge remote-tracking branch 'github/main' into asher/reconnection-with-screen

code-asher committed:
Remove unused fn

🤦

code-asher commented:
For posterity's sake, I made a mistake because I forgot waiting on the cond releases the lock; hotfix can be found here: https://github.com/coder/coder/pull/9094
